client/src/Features/MicrosoftLearn/MicrosoftLearnFilterCore.ts

```diff
@@ -0,0 +1,199 @@
+
+      keep = [...keep, ...productFilter.filter(f => !keep.includes(f) && !invisibleChildren.includes(f))];
+      return 'products='+keep.join('%2C');
+    }
+
+    let productUri = productFilter.length>0? getProductUri() : '';
+    let roleUri = roleFilter.length>0? 'roles=' + roleFilter.join('%2C') : '';
+    let typeUri = typeFilter.length>0? 'types=' + typeFilter.join('%2C') : '';
+    let levelUri = levelFilter.length>0? 'levels=' + levelFilter.join('%2C') : '';
+    let termsUri = searchTerm.length>0? 'terms='+searchTerm : '';
+    let expandedProductsUri = expandedProducts.length>0? 'expanded='+expandedProducts.join('%2C') : '';
+    let finalUri = [productUri, roleUri, levelUri, typeUri, termsUri, expandedProductsUri]
+                    .filter(s=>s.length!==0).join('&')            
+
+    let uri = window.location.protocol + "//" + window.location.host + window.location.pathname + '?' + finalUri;
+    window.history.pushState({path:uri},'',uri);       
+
+    return finalUri;
+}
+
+const getSearchTermFilteredLearnContent = (expressions: RegExp[], content: LearnContent[]): LearnContent[] => {
+    return content
+      .map(course => ({
+        course: course,
+        score: _.sumBy(
+          expressions,
+          singleExpression =>
+            scoreRegex(course.summary, singleExpression) + scoreRegex(course.title, singleExpression, 2)
+        )
+      }))
+      .filter(scouredCourse => scouredCourse.score > 0)
+      .sort((a, b) => b.score - a.score)
+      .map(scoredCourse => scoredCourse.course);
+}
+
+export const loadFiltersFromPath = (catalog: Catalog, productHierarchicalMap: Map<Product, Product[]>) => {
+    let searchParams = window.location.search
+    if(searchParams && searchParams.length>0){
+      let searchParamMap: Map<string, string[]> = new Map<string, string[]>();
+      searchParams.slice(1).split('&').forEach(s=>{
+        let [key, value] = s.split('=');
+        if(value){
+            searchParamMap.set(key, value.split('%2C'));
+        }
+      })
+
+      let _selectedFilters: Map<FilterType, string[]> = new Map<FilterType, string[]>();
+
+      let _expandedProducts = searchParamMap.get("expanded") || [];
+      _selectedFilters.set(FilterType.Role, searchParamMap.get('roles') || []);
+      _selectedFilters.set(FilterType.Level, searchParamMap.get('levels') || []);
+      _selectedFilters.set(FilterType.Type, searchParamMap.get('types') || []);
+
+      let setProducts = searchParamMap.get('products');
+      if(setProducts){
+        let parents = [...productHierarchicalMap.keys()];
+        let parenIds = parents.map(p => p.id);
+        
+        let selectedParents = setProducts.filter(p => parenIds.includes(p)).map(pId => catalog?.products.get(pId)!!); 
+        let addChildren = _.flatten(selectedParents.map(p => productHierarchicalMap.get(p)!!)).map(c=>c.id);
+        _selectedFilters.set(FilterType.Product,  [...setProducts, ...addChildren]);
```

Comment:
I believe this is the common code which I was talking about extracting into a separate method and making sure that `selectedFilters` are always in-line with it.